File: leetcode/golang/160_intersection_of_two_linked_lists.go
type ListNode struct {
    Val int
    Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
    tmpA, tmpB := headA, headB;
    for tmpA != tmpB {
        if tmpA == nil {
            tmpA = headB
        } else {
            tmpA = tmpA.Next
        }      
        if tmpB == nil {
            tmpB = headA
        } else {
            tmpB = tmpB.Next
        }
    }
    return tmpA
}

File: leetcode/c++/160_intersection_of_two_linked_lists.cpp
// https://leetcode.cn/problems/intersection-of-two-linked-lists/description

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *tmpA = headA, *tmpB = headB;
        while (tmpA != tmpB) {
            if (tmpA != NULL)
                tmpA = tmpA->next;
            else 
                tmpA = headB;
            if (tmpB != NULL)
                tmpB = tmpB->next;
            else 
                tmpB = headA;
        }
        return tmpA;
    }
};

// 时间复杂度为O(n+m)，空间复杂度为O(1)
// 还得是k神